### PR TITLE
use one copy for continuous area instead of loop copy

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
Signed-off-by: Chen Lihui <lihui.chen@sony.com>

### CODE REVIEW

**Review by fujitatomoya — APPROVED** [review]
lgtm with minor comment.

File: test/test_array_list.cpp [hunk 1]
```diff
@@ -374,6 +374,24 @@ TEST_F(ArrayListPreInitTest, remove_preserves_data_around_it) {
...
+  ret = rcutils_array_list_remove(&list, 0);
+  EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+
+  ret = rcutils_array_list_get_size(&list, &size);
+  EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+  EXPECT_EQ(size, (size_t)1);
+
+  ret = rcutils_array_list_get(&list, 0, &ret_data);
+  EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+  EXPECT_EQ((uint32_t)2, ret_data) << rcutils_get_error_string().str;
```
[the remark this answers is not in the document]
Comment (iuhilnehc-ynos) [reply test/test_array_list.cpp:394]: reasonable :+1:

File: test/test_array_list.cpp [hunk 2]
```diff
@@ -374,6 +374,31 @@ TEST_F(ArrayListPreInitTest, remove_preserves_data_around_it) {
...
   EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
   EXPECT_EQ((uint32_t)6, ret_data) << rcutils_get_error_string().str;
+
+  ret = rcutils_array_list_remove(&list, 2);
+  EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+
+  ret = rcutils_array_list_get_size(&list, &size);
+  EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+  EXPECT_EQ(size, (size_t)2);
+
```
Comment (clalancette) [inline test/test_array_list.cpp:384]: Can we also do `rcutils_array_list_get` for index 0 and 1 here, to ensure that they are 0 and 2, respectively?

**Review by clalancette — CHANGES_REQUESTED** [review]
I've got one small thing to fix.  Also, can you please run this test under `valgrind` just to ensure we aren't accessing any data out-of-bounds?  Once those two things are done, I'm happy to approve.

File: test/test_array_list.cpp (hunk 2 above)
Comment (iuhilnehc-ynos) [reply test/test_array_list.cpp:384]: Yes. Updated and rebased.

<details><summary> log for `valgrind ./test_array_list` </summary>
<p>

```shell
[----------] Global test environment tear-down
[==========] 35 tests from 2 test suites ran. (115 ms total)
[  PASSED  ] 35 tests.
==49168== 
==49168== HEAP SUMMARY:
==49168==     in use at exit: 0 bytes in 0 blocks
==49168==   total heap usage: 768 allocs, 768 frees, 165,188 bytes allocated
==49168== 
==49168== All heap blocks were freed -- no leaks are possible
==49168== 
==49168== For lists of detected and suppressed errors, rerun with: -s
==49168== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)


```

</p>
</details>

**Review by clalancette — APPROVED** [review]
Looks good to me now.  Thanks for iterating.  Will run CI on it next.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13047/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7995/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10771/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13029/)

**Comment by iuhilnehc-ynos** [comment]
Hi, @clalancette 
Do you think the test cases that failed in macOS and Windows are caused by this PR? If so, I'll investigate them on these platforms.

**Comment by clalancette** [comment]
> Do you think the test cases that failed in macOS and Windows are caused by this PR? If so, I'll investigate them on these platforms.

No, I don't think so.  Those tests have been flaky in the nightlies for a while.  This one just fell off of my radar, thanks for the ping.  I'm going to go ahead and merge this.

**Pull request merged**